accesshandler/pkg/providers/aws/ecs-shell-sso/provider.go

```diff
@@ -0,0 +1,119 @@
+	idStoreClient    *identitystore.Client
+	orgClient        *organizations.Client
+	awsAccountID     string
+
+	// the below fields are configured by gconfig
+
+	ecsClusterARN gconfig.StringValue
+	// sso instance
+	instanceARN gconfig.StringValue
+	// The globally unique identifier for the identity store, such as d-1234567890.
+	identityStoreID gconfig.StringValue
+	// The aws region where the identity store runs
+	ssoRegion gconfig.StringValue
+	ecsRegion gconfig.StringValue
+
+	// a role which can be assumed and has required sso and ecs permissions
+	ssoRoleArn gconfig.StringValue
+	ecsRoleArn gconfig.StringValue
+}
+
+func (p *Provider) Config() gconfig.Config {
+	return gconfig.Config{
+		gconfig.StringField("ecsClusterARN", &p.ecsClusterARN, "The ARN of the ECS Cluster to provision access to"),
+		gconfig.StringField("identityStoreId", &p.identityStoreID, "The AWS SSO Identity Store ID"),
+		gconfig.StringField("instanceArn", &p.instanceARN, "The AWS SSO Instance ARN"),
+		gconfig.StringField("ssoRoleArn", &p.ssoRoleArn, "The ARN of the AWS IAM Role with permission to administer SSO"),
+		gconfig.StringField("ecsRoleArn", &p.ecsRoleArn, "The ARN of the AWS IAM Role with permission to read ECS"),
+		gconfig.StringField("ssoRegion", &p.ssoRegion, "The region the AWS SSO instance is deployed to"),
+		gconfig.StringField("ecsRegion", &p.ecsRegion, "The region the ecs cluster instance is deployed to"),
+	}
+}
+
+// Init the provider.
+func (p *Provider) Init(ctx context.Context) error {
+	p.ssoCredentialCache = cfaws.NewAssumeRoleCredentialsCache(ctx, p.ssoRoleArn.Get(), cfaws.WithRoleSessionName("accesshandler-ecs-shell-sso"))
+	p.ecsCredentialCache = cfaws.NewAssumeRoleCredentialsCache(ctx, p.ecsRoleArn.Get(), cfaws.WithRoleSessionName("accesshandler-ecs-shell-sso"))
+
+	ssoCfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(p.ssoRegion.Get()), config.WithCredentialsProvider(p.ssoCredentialCache))
+	if err != nil {
+		return err
+	}
+	ecsCfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(p.ecsRegion.Get()), config.WithCredentialsProvider(p.ecsCredentialCache))
+	if err != nil {
+		return err
+	}
+
+	p.cloudtrailClient = cloudtrail.NewFromConfig(ecsCfg)
+	p.ssmClient = ssm.NewFromConfig(ecsCfg)
+	p.ecsClient = ecs.NewFromConfig(ecsCfg)
+	p.ssoClient = ssoadmin.NewFromConfig(ssoCfg)
+	p.orgClient = organizations.NewFromConfig(ssoCfg)
+	p.idStoreClient = identitystore.NewFromConfig(ssoCfg)
+	p.iamClient = iam.NewFromConfig(ssoCfg)
+	stsClient := sts.NewFromConfig(ecsCfg)
+	res, err := stsClient.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
+	if err != nil {
+		return err
+	}
+	if res.Account == nil {
+		return errors.New("aws accountID was nil in sts get caller id response")
+	}
+	p.awsAccountID = *res.Account
+
+	//check to see if cluster is running and has exec enabled
+	clusters, err := p.ecsClient.DescribeClusters(ctx, &ecs.DescribeClustersInput{Clusters: []string{p.ecsClusterARN.Get()}})
```

Comment:
@chrnorm There is a check here that can fail if the state of the cluster changes.
The problem I see is that because this happens during Init, if it fails it will prevent the Access handler from ever starting, causing a complete outage of Approvals.

While its obvious that grants will fail if this is not enabled. I think this shows that we need to update our provider initialisation to make individual providers as failed rather than failing to start the access handler service at all.

As a start, I will shift this check into a validation that runs during provider setup, and we can ticket the work separately to improve the stability of the access handler

Reply:
For this release I might just leave this as is to get it out